### ISSUE TITLE
Scale format-value event crash (invalid pointer)

### ISSUE THREAD

**Issue by Dankr4d** [StefanSalewski/gintro#161]
Following code causes an `invalid pointer` crash (tested with refc, arc and orc garbage collector) .
I've tested it with gintro 0.9.3 and 0.9.4, using Nim version 1.4.8.

**Example:**
```nim
import gintro/[gtk, gobject, gio]

proc onScaleFormatValue(self: Scale, value: cdouble): string =
  return "My formated string"

proc appActivate(app: Application) =
  let window = newApplicationWindow(app)
  window.defaultSize = (250, 50)
  let adj = newAdjustment(1, 1, 4, 1, 1, 1)
  let scale = newScale(Orientation.horizontal, adj)
  scale.digits = 0
  window.add(scale)
  scale.connect("format-value", onScaleFormatValue)
  window.showAll()

proc main() =
  let app = newApplication("org.gtk.example")
  connect(app, "activate", appActivate)
  discard app.run()

main()
```

**Output:**
```bash
Hint: /home/dankrad/Desktop/testsliderformat  [Exec]
munmap_chunk(): invalid pointer
Traceback (most recent call last)
/home/dankrad/Desktop/testsliderformat.nim(21) testsliderformat
/home/dankrad/Desktop/testsliderformat.nim(19) main
/home/dankrad/.nimble/pkgs/gintro-0.9.3/gintro/gio.nim(31002) run
/opt/Nim/lib/core/macros.nim(559) connect_for_signal_cdecl_activate2
/home/dankrad/Desktop/testsliderformat.nim(14) appActivate
/home/dankrad/.nimble/pkgs/gintro-0.9.3/gintro/gtk.nim(2991) showAll
SIGABRT: Abnormal termination.
Error: execution of an external program failed: '/home/dankrad/Desktop/testsliderformat '
```

**Comment by StefanSalewski** [comment]
The crash is not really surprising.

https://developer.gnome.org/gtk3/stable/GtkScale.html#GtkScale-format-value

The user function returns a char* which is then freed by glib. That is a very special case, most user functions return nothing or a bool value only.

Note that this signal has vanished in GTK4. So we need a special handling for this signal in GTK3 and for GTK4 we would need a macro similar like the idleAdd() macro in gimplgob.nim.

**Comment by StefanSalewski** [comment]
For the returned cstring the fix seems to be easy, we just have to call g_strdup() on it so that glib can free it with g_free. Will provide the fix tomorrow. Well only for GTK3, for GTK4 we need a macro.

**Comment by StefanSalewski** [comment]
Should be fixed for GTK3. Try with

nimble uninstall gintro
nimble install gintro@#head

For GTK4 we have to create a macro instead. And there seems to exist a few connect() user functions which returns gobject structs, for which I am unsure if they work. But I will delay that investigation until someone really uses one of them.

I think I will create the GTK4 macro in the next few days, as it should be easy, and we have above test code. But it is ugly that we need so many separate macros, we should try to find a better way.

Note that in your code above you can use Nim float instead of cdouble. For the current Nim implementation float is identical to cdouble, and generally gintro bindings should always work with float.

**Comment by Dankr4d** [comment]
Everything works and thank you for clearing up, that float and cdouble is identical in nim. I'll fix it.

For those who are connecting to low level events with `Builder`, you need to duplicate the cstring with `g_strdup`, like this:
```nim
proc onScaleFormatValue00(self: ptr Scale00, value: float): cstring {.exportc, cdecl, dynlib.} =
  return g_strdup("My formatted string")
```

Thank you!

**Comment by StefanSalewski** [comment]
Should be fixed in v0.9.4 now.
Will close this issue, please reopen if necessary.
